lib/ecto/changeset.ex

```diff
@@ -1498,17 +1497,16 @@ defmodule Ecto.Changeset do
       iex> {:error, changeset} = apply_action(changeset, :update)
       %Ecto.Changeset{action: :update}
   """
-  @spec apply_action(t, :insert | :update | :delete | :replace) ::
-          {:ok, Ecto.Schema.t() | data} | {:error, t}
-  def apply_action(%Changeset{} = changeset, action) when action in @actions do
+  @spec apply_action(t, atom) :: {:ok, Ecto.Schema.t() | data} | {:error, t}
```

Comment:
```suggestion
  @spec apply_action(t, action) :: {:ok, Ecto.Schema.t() | data} | {:error, t}
```

Reply:
Thanks for the suggestion. This seems like it would be functionally equivalent (although it would allow for the possibility to rollaback the change to allow any atom in actions.) Either way, I'm happy to add your suggested change.

Reply:
I think I will go with atom just because `action` has both `nil` and `:ignore` that we wouldn't place here. It is kind of a weak argument, because those can be passed as `atom` anyway, but maybe as atom it calls less attention to them.